### PR TITLE
Logger with streams implementation

### PR DESCRIPTION
**Pull request by JDuchniewicz**

### CODE REVIEW

File: PolyEngine/Core/Core.vcxproj [hunk 1]
```diff
@@ -179,6 +179,7 @@
     <ClInclude Include="Src\Matrix.hpp" />
     <ClInclude Include="Src\Optional.hpp" />
     <ClInclude Include="Src\ObjectLifetimeHelpers.hpp" />
+    <ClInclude Include="Src\OutputStream.hpp" />
```
Comment (MuniuDev) [inline PolyEngine/Core/Core.vcxproj]: please add this file to CMAkeLists.txt as well.

File: PolyEngine/Core/Src/Logger.hpp [hunk 2]
```diff
@@ -23,8 +25,10 @@ namespace Poly
 	inline void sprint(std::ostream& stream, eLogLevel level,
 		const std::string& fmt) {
 		(void)level;  // suppress unused warning
-		stream << fmt << std::endl;
+		stream << fmt.c_str();
+		stream << String("\n");
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/Logger.hpp]: This line is unnecessary. Remove it.

File: PolyEngine/Core/Src/Logger.hpp [hunk 3]
```diff
@@ -40,18 +44,43 @@ namespace Poly
 		const T& head, Args&&... tail) {
 		constexpr char marker[] = "{}";
 		std::string::size_type format_marker_pos = fmt.find(marker);
-		stream << fmt.substr(0, format_marker_pos);
+		stream << String(fmt.substr(0, format_marker_pos).c_str());
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/Logger.hpp]: Leave std::string uasge here, we will refactor it later.

File: PolyEngine/Core/Src/Logger.hpp [hunk 4]
```diff
@@ -40,18 +44,43 @@ namespace Poly
...
 
 	class CORE_DLLEXPORT Console : public BaseObject<> {
 	public:
+		Console() : Ostream(new std::ostream(std::cout.rdbuf())){} // this is a hack :v
+
+		//thoughts
+		//cannot assign std::cout to ostream via copy
+		//what to do with sentinel function outside of class
+
+		//handle checking if new is the same as old one
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/Logger.hpp]: Remove above comments. They are irrelevant now.

File: PolyEngine/Core/Src/Logger.hpp [hunk 5]
```diff
@@ -40,18 +44,43 @@ namespace Poly
...
 				tail...);
 		}
 		else
-			stream << std::endl;
+			stream << String("\n");
 	}
 
 	class CORE_DLLEXPORT Console : public BaseObject<> {
 	public:
+		Console() : Ostream(new std::ostream(std::cout.rdbuf())){} // this is a hack :v
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/Logger.hpp]: This is not hack. remove this comment. Also, codding style, put space before first bracket.

File: PolyEngine/Core/Src/Logger.hpp [hunk 6]
```diff
@@ -40,18 +44,43 @@ namespace Poly
...
+		void RegisterStream(Args&&... args)
+		{
+			constexpr bool isStream = std::is_base_of<OutputStream, S>::value; // Strange workaround to STATIC_ASSERTE macro on MSVC
+			STATIC_ASSERTE(isStream, "Provided value is not stream!");
+			UnregisterStream();
+			CurrentStream = std::make_unique<S>(std::forward<Args>(args)...);
+			Ostream.reset( new std::ostream(CurrentStream.get()));
+		}
+
+		void UnregisterStream() {
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/Logger.hpp]: Codding style. Move bracket to new line.

File: PolyEngine/Core/Src/OutputStream.hpp [hunk 7]
```diff
@@ -0,0 +1,65 @@
+﻿#pragma once
+
+#include "Defines.hpp"
+#include <ostream>
+
+namespace Poly {
+
+	class OutputStream : public BaseObject<>, public std::streambuf
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/OutputStream.hpp]: Create cpp file and move all (multiline) method implementations to it.

File: PolyEngine/Core/Src/OutputStream.hpp [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+		{
+			char tab[2];
+			tab[0] = c;
+			tab[1] = '\0';
+			Append(String(tab));
+			return c;
+		}
+	};
+
+	class FileOutputStream : public OutputStream {
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/OutputStream.hpp]: CS. Bracket should be moved to new line.

File: PolyEngine/Core/Src/OutputStream.hpp [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+	public:
+		FileOutputStream(const char* name) {
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/OutputStream.hpp]: CS. Bracket should be moved to new line.

File: PolyEngine/Core/Src/OutputStream.hpp [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+	public:
+		FileOutputStream(const char* name) { 
+#if defined(_WIN32)
+			freopen_s(&file, name, "w", stdout); 
+#else
+			file = freopen(name, "w", stdout);
+#endif
+			setvbuf(stdout, NULL, _IONBF, 0);
+		}
+		FileOutputStream& operator=(FileOutputStream&& rhs) {
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/OutputStream.hpp]: CS. Put new line between implementations. Bracket should be moved to new line.

File: PolyEngine/Core/Src/OutputStream.hpp [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+			Append(String(tab));
+			return c;
+		}
+	};
+
+	class FileOutputStream : public OutputStream {
+	public:
+		FileOutputStream(const char* name) { 
+#if defined(_WIN32)
+			freopen_s(&file, name, "w", stdout);
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/OutputStream.hpp]: This is the reason we're not geting output to std::cout. Use fopen instead.

File: PolyEngine/Core/Src/OutputStream.hpp [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+	class FileOutputStream : public OutputStream {
+	public:
+		FileOutputStream(const char* name) { 
+#if defined(_WIN32)
+			freopen_s(&file, name, "w", stdout); 
+#else
+			file = freopen(name, "w", stdout);
+#endif
+			setvbuf(stdout, NULL, _IONBF, 0);
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/OutputStream.hpp]: Provide error message if file open failed. Change to 
```
if(file)
    setvbuf(file, nullptr, _IONBF, 0);
```

File: PolyEngine/Core/Src/OutputStream.hpp [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+		}
+		FileOutputStream& operator=(FileOutputStream&& rhs) {
+			EnsureFileClosed();
+			file = rhs.file;
+			rhs.file = nullptr;
+			return *this;	
+		}
+		FileOutputStream(FileOutputStream&& rhs) {
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/OutputStream.hpp]: CS. Put new line between implementations. Bracket should be moved to new line.

File: PolyEngine/Core/Src/OutputStream.hpp [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+			*this = std::move(rhs);
+		}
+		FileOutputStream() = default;
+		~FileOutputStream() { EnsureFileClosed(); }
+		void OnUnregister() override { EnsureFileClosed(); }
+		void EnsureFileClosed() {
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/OutputStream.hpp]: CS. Put new line between implementations. Bracket should be moved to new line.

File: PolyEngine/Core/Src/OutputStream.hpp [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+			if (file) { fclose(file); file = nullptr; }
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/OutputStream.hpp]: CS. Brackets and statements should be moved to new lines.

File: PolyEngine/Core/Src/OutputStream.hpp [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+		}
+		bool IsFileOpen() const { return file != nullptr; }
+		void Append(String&& data) override {
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/OutputStream.hpp]: CS. Put new line between implementations. Bracket should be moved to new line.

File: PolyEngine/Core/Src/OutputStream.hpp [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+			fwrite(data.GetCStr(), sizeof(char), data.GetLength(), file);
+		}
+	private:
+		FILE* file = nullptr;
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/OutputStream.hpp]: Rename to `FileHandle`

File: PolyEngine/Standalone/Src/Main.cpp [hunk 18]
```diff
@@ -3,15 +3,29 @@
...
-using CreateGameFunc = Poly::IGame* (void);
+extern "C"
+{
+	using CreateRenderingDeviceFunc = Poly::IRenderingDevice* (SDL_Window*, const Poly::ScreenSize&);
+	using CreateGameFunc = Poly::IGame* (void);
+}
 
 void HandleWindowEvent(const SDL_WindowEvent& windowEvent);
 
+	class FileAndCoutStream : public Poly::FileOutputStream {
```
Comment (MuniuDev) [inline PolyEngine/Standalone/Src/Main.cpp]: CS. Fix indentation. Move bracket to new line.

File: PolyEngine/Core/Src/OutputStream.hpp [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+#include "Defines.hpp"
+#include <ostream>
+
+namespace Poly {
+
+	class OutputStream : public BaseObject<>, public std::streambuf
+	{
+	public:
+		virtual void OnUnregister() {}
+		virtual void Append(String&& str) = 0;
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/OutputStream.hpp]: Let's refactor this method to be `Append(const char*)`

File: PolyEngine/Core/Src/Logger.hpp (hunk 2 above)
Comment (JDuchniewicz) [reply PolyEngine/Core/Src/Logger.hpp]: I leave it here, because otherwise we get everything in one line, just checked it

Comment (JDuchniewicz) [reply PolyEngine/Core/Src/Logger.hpp]: I replaced it with std::endl

File: PolyEngine/Core/Src/OutputStream.hpp [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+		std::streamsize xsputn(const char_type* s, std::streamsize n) override final
+		{
+			UNUSED(n);
+			std::string str(s);
+			size_t len = str.length();
+			Append(s);
+			return len;
+		}
+
+		int_type overflow(int_type c) override final
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/OutputStream.hpp]: Move `overflow` and `xsputn` implementation to cpp.

File: PolyEngine/Core/Src/Logger.hpp [hunk 21]
```diff
@@ -50,8 +53,30 @@ namespace Poly
...
+		void RegisterStream(Args&&... args)
+		{
+			constexpr bool isStream = std::is_base_of<OutputStream, S>::value; // Strange workaround to STATIC_ASSERTE macro on MSVC
+			STATIC_ASSERTE(isStream, "Provided value is not stream!");
+			UnregisterStream();
+			CurrentStream = std::make_unique<S>(std::forward<Args>(args)...);
+			Ostream.reset( new std::ostream(CurrentStream.get()));
+		}
+
+		void UnregisterStream()
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/Logger.hpp]: Let's remove this function, and pase its cuntent in line 66.

**Pull request merged**